qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py

```diff
@@ -0,0 +1,756 @@
+
+        # 2) Find which delays are active during which time windows, where a time window
+        # is defined as (breakpoints[i], breakpoints[i + 1])
+        active_delays = {}  # {window: [group1, group2]} where each group = (index1, index2, ..)
+        op_map = {}  # {(qubit_index, window): delay operation as DAGOpNode}
+        windows = list(zip(breakpoints[:-1], breakpoints[1:]))
+        for window in windows:
+            active = []
+
+            # check which delays are active during the window
+            # this could be e.g. [0, 1, 2, 5, 6, 9]
+            for index, delays in all_delays.items():
+                for delay in delays:
+                    # since the windows are atomic, we have three cases:
+                    # (1) the delay starts at the time window, (2) it ends with it, (3) it contains it
+                    if delay.start <= window[0] and delay.end >= window[1]:
+                        active.append(index)
+                        delay.add_window(window)
+                        op_map[(index, window)] = delay
+
+            # check which are adjacent
+            # on a linear topology, we would get [[0, 1, 2], [5, 6], [9]]
+            visited = defaultdict(lambda: False)
+            grouped = []
+            for start_index in active:
+                if visited[start_index]:
+                    continue
+
+                active_neighbors = {start_index}
+                _dfs(start_index, self._coupling_map, active_neighbors, active)
+
+                for index in active_neighbors:
+                    visited[index] = True
+
+                group = tuple(sorted(active_neighbors))  # must be sorted to merge later
+                grouped.append(group)
+
+            if validate:
+                # sanity check: groups must be disjoint
+                for i, g1 in enumerate(grouped):
+                    for g2 in grouped[i + 1 :]:
+                        if len(set(g1).intersection(g2)) > 0:
+                            raise RuntimeError(f"Groups not disjoint: {g1} and {g2}.")
+
+            active_delays[window] = grouped
+
+        # 3) Merge time-adjacent active delays
+        merged_delays = []
+        open_groups = {}
+        for window in windows:
+            next_groups = active_delays[window]
+
+            # check which opened groups are still active in this window
+            next_open_groups = {}
+            for open_group, delay in open_groups.items():
+                if open_group in next_groups:
+                    # extend the delay operation and remove a breakpoint
+                    delay.end = window[1]
+                    for op in delay.ops:
+                        op.breakpoints.remove(window[0])
+
+                    # TODO this is a sanity check
+                    involved_ops = {op_map[(index, window)] for index in open_group}
+                    if len(involved_ops.difference(delay.ops)) > 0:
+                        raise RuntimeError("I cannot think")
```

Comment:
You are thinking too classically ;-)
Make it an assertion or part of a validation block